File: code/go/lc-api.go
package main

import (
    "context"
    "database/sql"
    "encoding/json"
    "fmt"
    "log"
    "net/http"
    "os"

    "cloud.google.com/go/pubsub"
    "github.com/gin-gonic/gin"
    _ "github.com/go-sql-driver/mysql"
    "github.com/jmoiron/sqlx"
)

type Person struct {
    ID       string `json:"id" binding:"required"`
    Name     string `json:"name" binding:"required"`
    LastName string `json:"last_name" binding:"required"`
    FieldA   string `json:"field_a"`
    FieldB   string `json:"field_b"`
    FieldC   string `json:"field_c"`
}

const insertSQL = `INSERT INTO lc_table (ID, Name, LastName, FieldA, FieldB, FieldC) VALUES (?, ?, ?, ?, ?, ?)`

var (
    db             *sqlx.DB
    pubSubClient   *pubsub.Client
    pubSubTopic    *pubsub.Topic
    dbUser         string
    dbPass         string
    dbName         string
    dbInstance     string
    pubSubTopicID  string
    connectionName string
)

func main() {
    var err error

    // Load environment variables
    dbUser = os.Getenv("DB_USER")
    dbName = os.Getenv("DB_DATABASE")
    dbInstance = os.Getenv("DB_INSTANCE")
    pubSubTopicID = os.Getenv("PUBSUB_TOPIC")
    connectionName = os.Getenv("CLOUDSQL_CONNECTION_NAME")

    // Fetch DB password from Secret Manager (pseudo-code)
    dbPass = fetchSecret(os.Getenv("CLOUDSQL_PASSWD"))

    // Set up CloudSQL connection
    dataSourceName := fmt.Sprintf("%s:%s@unix(/cloudsql/%s)/%s", dbUser, dbPass, connectionName, dbName)
    db, err = sqlx.Open("mysql", dataSourceName)
    if err != nil {
        log.Fatalf("Could not connect to the database: %v", err)
    }
    defer db.Close()

    // Set up Pub/Sub client
    ctx := context.Background()
    projectID := os.Getenv("PROJECT_ID")
    if projectID == "" {
        log.Fatal("Environment variable PROJECT_ID is not set")
    }
    pubSubClient, err = pubsub.NewClient(ctx, projectID)
    if err != nil {
        log.Fatalf("Failed to create PubSub client: %v", err)
    }
    defer pubSubClient.Close()

    pubSubTopic = pubSubClient.Topic(pubSubTopicID)

    // Set up Gin router
    router := gin.Default()
    router.POST("/ingest", handleIngest)
    router.GET("/retrieve", handleRetrieve)

    // Run the server
    if err := router.Run(":8080"); err != nil {
        log.Fatal("Unable to start server: ", err)
    }
}

func fetchSecret(secretName string) string {
    // This function should be implemented to fetch the secret from GCP Secret Manager
    // For the purpose of this example, it will return a placeholder value
    return "your-secret-value"
}

func handleIngest(c *gin.Context) {
    var person Person
    if err := c.ShouldBindJSON(&person); err != nil {
        c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
        return
    }

    // Insert into CloudSQL
    _, err := db.Exec(insertSQL, person.ID, person.Name, person.LastName, person.FieldA, person.FieldB, person.FieldC)
    if err != nil {
        c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to insert data into database"})
        return
    }

    // Publish to Pub/Sub
    personBytes, err := json.Marshal(person)
    if err != nil {
        c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to marshal data"})
        return
    }
    result := pubSubTopic.Publish(context.Background(), &pubsub.Message{Data: personBytes})

    // Block until the result is returned and a server-generated ID is returned for the published message
    _, err = result.Get(context.Background())
    if err != nil {
        c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to publish to Pub/Sub"})
        return
    }

    c.JSON(http.StatusOK, gin.H{"message": "Data ingested successfully"})
}

func handleRetrieve(c *gin.Context) {
    field := c.Query("field")
    criteria := c.Query("criteria")

    if field == "" || criteria == "" {
        c.JSON(http.StatusBadRequest, gin.H{"error": "Missing field or criteria parameters"})
        return
    }

    // Perform a search in the database
    var results []Person
    query := fmt.Sprintf("SELECT * FROM lc_table WHERE %s = ?", field)
    err := db.Select(&results, query, criteria)
    if err != nil {
        if err == sql.ErrNoRows {
            c.JSON(http.StatusOK, gin.H{"message": "No data found"})
        } else {
            c.JSON(http.StatusInternalServerError, gin.H{"error": "Error during database query"})
        }
        return
    }

    c.JSON(http.StatusOK, gin.H{"data": results})
}

File: code/python/lc-api.py
import os
import base64
import json
from flask import Flask, request
from google.cloud import pubsub_v1
from google.cloud import secretmanager
from google.cloud import sql

app = Flask(__name__)

# Define environment variables
project_id = os.environ.get('PROJECT_ID')
cloudsql_instance_name = os.environ.get('CLOUDSQL_INSTANCE_NAME')
database_name = os.environ.get('DATABASE_NAME')
database_schema = os.environ.get('DATABASE_SCHEMA')
database_username = os.environ.get('DATABASE_USERNAME')

# Create a Cloud SQL connection
db = sql.connector.connect(
    user=database_username,
    password=secretmanager.SecretManagerServiceClient().access_secret_version(
        request={"name": "projects/{project_id}/secrets/[SECRET-NAME]/versions/latest"}
    ).payload.data.decode("UTF-8"),
    host=f"/cloudsql/{project_id}:{cloudsql_instance_name}",
    database=database_name
)

# Create a Pub/Sub publisher
publisher = pubsub_v1.PublisherClient()
topic_path = publisher.topic_path(project_id, 'lc-bq-update')

@app.route('/ingestion', methods=['PUT'])
def ingestion():
    data = request.get_json()

    # Write data to Cloud SQL
    cursor = db.cursor()
    query = f"INSERT INTO {database_schema} (ID, Name, LastName, FieldA, FieldB, FieldC) " \
            f"VALUES('{data['ID']}', '{data['Name']}', '{data['LastName']}', '{data['FieldA']}', " \
            f"'{data['FieldB']}', '{data['FieldC']}')"
    cursor.execute(query)
    db.commit()

    # Publish data to Pub/Sub
    message = json.dumps(data).encode('utf-8')
    future = publisher.publish(topic_path, data=message)
    future.result()

    return 'Data ingested and published to Pub/Sub'

@app.route('/query', methods=['GET'])
def query():
    what = request.args.get('what')
    where = request.args.get('where')

    # Query data from Cloud SQL
    cursor = db.cursor()
    query = f"SELECT * FROM {database_schema} WHERE {where} = '{what}'"
    cursor.execute(query)
    results = cursor.fetchall()

    response = []
    for row in results:
        response.append({
            'ID': row[0],
            'Name': row[1],
            'LastName': row[2],
            'FieldA': row[3],
            'FieldB': row[4],
            'FieldC': row[5]
        })

    return json.dumps(response)

if __name__ == '__main__':
    app.run()